File: fsharp/luhn/Luhn.fs
module Luhn

open System

let luhnDigit i x =
    if i % 2 = 0 then
        x
    else
        let x' = x * 2
        if x' > 9 then x' - 9 else x'

let valid (number : string) =
    let trimmed = number |> Seq.filter ((<>) ' ')

    if Seq.length trimmed <= 1 then
        false
    elif not (trimmed |> Seq.forall Char.IsDigit) then
        false
    else
        trimmed
        |> Seq.filter Char.IsDigit
        |> Seq.map (Char.GetNumericValue >> int)
        |> Seq.rev
        |> Seq.mapi luhnDigit
        |> Seq.sum
        |> (fun x -> x % 10)
        |> (=) 0

File: go/luhn/luhn.go
package luhn

func Valid(input string) bool {
	sum := 0
	length := 0

	for i := len(input) - 1; i >= 0; i-- {
		if input[i] == ' ' {
			continue
		}
		v := int(input[i] - '0')
		if v < 0 || v > 9 {
			return false
		}
		length++
		if length%2 == 0 {
			v *= 2
			if v > 9 {
				v -= 9
			}
		}
		sum += v
	}

	return length > 1 && sum%10 == 0
}
